nexus/src/app/saga.rs

```diff
@@ -56,70 +85,91 @@ pub(crate) fn create_saga_dag<N: NexusSaga>(
-        &self,
-        opctx: &OpContext,
-        pagparams: &DataPageParams<'_, Uuid>,
-    ) -> ListResult<nexus_types::internal_api::views::Saga> {
-        // The endpoint we're serving only supports `ScanById`, which only
-        // supports an ascending scan.
-        bail_unless!(
-            pagparams.direction == dropshot::PaginationOrder::Ascending
-        );
-        opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
-        let marker = pagparams.marker.map(|s| SagaId::from(*s));
-        let saga_list = self
-            .sec_client
-            .saga_list(marker, pagparams.limit)
-            .await
-            .into_iter()
-            .map(nexus_types::internal_api::views::Saga::from)
-            .map(Ok);
-        Ok(futures::stream::iter(saga_list).boxed())
-    }
+/// External handle to a self-contained subsystem for kicking off sagas
+///
+/// Note that Steno provides its own interface for kicking off sagas.  This one
+/// is a thin wrapper around it.  This one exists to layer Nexus-specific
+/// behavior on top of Steno's (e.g., error conversion).
+pub(crate) struct SagaExecutor {
+    sec_client: Arc<steno::SecClient>,
+    log: slog::Logger,
+    nexus: OnceLock<Arc<Nexus>>,
+}
 
-    pub(crate) async fn saga_get(
-        &self,
-        opctx: &OpContext,
-        id: Uuid,
-    ) -> LookupResult<nexus_types::internal_api::views::Saga> {
-        opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
-        self.sec_client
-            .saga_get(SagaId::from(id))
-            .await
-            .map(nexus_types::internal_api::views::Saga::from)
-            .map(Ok)
-            .map_err(|_: ()| {
-                Error::not_found_by_id(ResourceType::SagaDbg, &id)
-            })?
+impl SagaExecutor {
+    pub(crate) fn new(
+        sec_client: Arc<steno::SecClient>,
+        log: slog::Logger,
+    ) -> SagaExecutor {
+        SagaExecutor { sec_client, log, nexus: OnceLock::new() }
     }
 
-    pub(crate) async fn create_runnable_saga(
-        self: &Arc<Self>,
-        dag: SagaDag,
-    ) -> Result<RunnableSaga, Error> {
-        // Construct the context necessary to execute this saga.
-        let saga_id = SagaId(Uuid::new_v4());
+    // This is a little gross.  We want to hang the SagaExecutor off of Nexus,
+    // but we also need to refer to Nexus, which thus can't exist when
+    // SagaExecutor is constructed.  So we have the caller hand it to us after
+    // initialization.
+    //
+    // This isn't as statically verifiable as we'd normally like.  But it's only
+    // one call site, it does fail cleanly if someone tries to use
+    // `SagaExecutor` before this has been set, and the result is much cleaner
+    // for all the other users of `SagaExecutor`.
+    pub(crate) fn set_nexus(&self, nexus: Arc<Nexus>) {
+        self.nexus.set(nexus).unwrap_or_else(|_| {
+            panic!("concurrent initialization of SagaExecutor")
+        })
+    }
```

Comment:
So this will panic if called twice, which is kinda spooky as undocumented behavior.

WDYT about adding a `# panic` documentation section, or attempting to ignore the second invocation?

(This is unlikely, just trying to help grind down sharp edges)

Reply:
I'm happy to add a "Panics" section to the docs.  It doesn't seem worth it to me to try to handle this better but I'm open to ideas there.

Reply:
I documented this and also tweaked the message.  This will panic if the function is called twice in the lifetime of the object.

We _could_ make this a noop by just ignoring the error, and the assumption would be: that means somebody else set the value, and it must be a valid `Arc<Nexus>, so whatever.  But I feel like this case almost certainly reflects a mistake, and possibly one that will result in very confusing behavior later, so that's why I want it to panic.

Reply:
I'm okay with documenting the panic! That seems like the most we should invest in this otherwise pretty unlikely behavior, for now